### PR TITLE
GET /api/articles (queries)

### PR DESCRIPTION
**Pull request by SyedaKausar**

### CODE REVIEW

File: models/articles.models.js [hunk 1]
```diff
@@ -51,15 +32,58 @@ exports.fetchArticleByIdToPatch = (id, incrementVotes) => {
...
-  FROM articles
-  LEFT JOIN comments ON articles.article_id = comments.article_id  
-  GROUP BY articles.article_id
-  ORDER BY created_at DESC`;
-  return db.query(queryStr).then((body) => {
-    return body.rows;
-  });
+exports.fetchAllArticles = (sort_by = "created_at", order = "DESC", topic) => {
+  return db
+    .query(`SELECT topic FROM articles GROUP BY topic`)
```
Comment (hevdevs) [inline models/articles.models.js]: We really want to SELECT slug FROM topics instead of selecting topic from articles here, as if a new topic is added to your db it will appear in the topics table first (and would still be a valid option even if it had no relevant articles yet)

File: __tests__/app.test.js [hunk 2]
```diff
@@ -289,3 +289,139 @@ describe("POST /api/articles/:article_id/comments", () => {
...
+          },
+        ]);
+      });
+  });
+  test("status 404: not found when topic is not in database", () => {
+    return request(app)
+      .get("/api/articles?topic=no")
+      .expect(404)
+      .then(({ body }) => {
+        console.log(body);
```
Comment (hevdevs) [inline __tests__/app.test.js]: remove cl before merge

File: __tests__/app.test.js [hunk 3]
```diff
@@ -289,3 +289,139 @@ describe("POST /api/articles/:article_id/comments", () => {
...
+        expect(body.msg).toBe("not found");
+      });
+  });
+  test("status 400: bad req when sort_by coloumn doesnt exist", () => {
+    return request(app)
+      .get("/api/articles?sort_by=notanumber&order=asc")
+      .expect(400)
+      .then(({ body }) => {
+        console.log(body);
```
Comment (hevdevs) [inline __tests__/app.test.js]: remove cl before merge

File: __tests__/app.test.js [hunk 4]
```diff
@@ -289,3 +289,139 @@ describe("POST /api/articles/:article_id/comments", () => {
...
+        expect(body.msg).toBe("bad request");
+      });
+  });
+  test("status 400: bad req when order is not asc or desc", () => {
+    return request(app)
+      .get("/api/articles?sort_by=no&order=notasc")
+      .expect(400)
+      .then(({ body }) => {
+        console.log(body);
```
Comment (hevdevs) [inline __tests__/app.test.js]: remove cl before merge

File: __tests__/app.test.js [hunk 5]
```diff
@@ -289,3 +289,139 @@ describe("POST /api/articles/:article_id/comments", () => {
...
+        expect(body.msg).toBe("bad request");
+      });
+  });
+  test("status 404: bad req when topic is not in database", () => {
+    return request(app)
+      .get("/api/articles?topic=no")
+      .expect(404)
+      .then(({ body }) => {
+        console.log(body);
```
Comment (hevdevs) [inline __tests__/app.test.js]: remove cl before merge

File: models/articles.models.js [hunk 6]
```diff
@@ -51,15 +32,58 @@ exports.fetchArticleByIdToPatch = (id, incrementVotes) => {
...
+
+      if (sort_by) {
+        if (validSortBy.includes(sort_by)) {
+          queryStr += ` ORDER BY ${sort_by} ${orderStr}`;
+        } else {
+          console.log(sort_by);
+          return Promise.reject({ status: 400, msg: "bad request" });
+        }
+      }
+      console.log(queryStr);
```
Comment (hevdevs) [inline models/articles.models.js]: remove cl before merge

File: models/articles.models.js [hunk 7]
```diff
@@ -51,15 +32,58 @@ exports.fetchArticleByIdToPatch = (id, incrementVotes) => {
...
+      
+     `;
+      if (topic) {
+        if (topicArray.includes(topic)) {
+          queryStr += ` WHERE articles.topic = '${topic}'`;
+        } else {
+          return Promise.reject({ status: 404, msg: "not found" });
+        }
+      }
+      // } else {
```
Comment (hevdevs) [inline models/articles.models.js:64]: remove comments if not needed before merge

**Review by hevdevs — CHANGES_REQUESTED** [review]
Few suggestions here, you also want a 200 status test for when a topic exists but has no articles, in which case we want the response to be an empty array

**Review by hevdevs — APPROVED** [review]
